wallet: fix crash on loading descriptor wallet containing legacy key type entries

theStack opened the pull request:
Loading a descriptor wallet currently leads to a segfault if a legacy key type entry is present that can be deserialized successfully and needs SPKman-interaction. To reproduce with a "cscript" entry (see second commit for details):

```
$ ./src/bitcoin-cli createwallet crashme
$ ./src/bitcoin-cli unloadwallet crashme
$ sqlite3 ~/.bitcoin/wallets/crashme/wallet.dat
SQLite version 3.38.2 2022-03-26 13:51:10
Enter ".help" for usage hints.
sqlite> INSERT INTO main VALUES(x'07637363726970740000000000000000000000000000000000000000', x'00');
$ ./src/bitcoin-cli loadwallet crashme

--- bitcoind output: ---
2022-11-06T13:51:01Z Using SQLite Version 3.38.2
2022-11-06T13:51:01Z Using wallet /home/honey/.bitcoin/wallets/crashme
2022-11-06T13:51:01Z init message: Loading wallet…
2022-11-06T13:51:01Z [crashme] Wallet file version = 10500, last client version = 249900

Segmentation fault (core dumped)
```

Background: In the wallet key-value-loading routine, most legacy type entries require a `LegacyScriptPubKeyMan` instance after successful deserialization. On a descriptor wallet, creating that (via method `GetOrCreateLegacyScriptPubKeyMan`) fails and then leads to a null-pointer dereference crash. E.g. for CSCRIPT: https://github.com/bitcoin/bitcoin/blob/50422b770a40f5fa964201d1e99fd6b5dc1653ca/src/wallet/walletdb.cpp#L589-L594


~~This PR fixes this by simply ignoring legacy entries if the wallet flags indicate that we have a descriptor wallet. The second commits adds a regression test to the descriptor wallet's functional test (fortunately Python includes sqlite3 support in the standard library).~~

~~Probably it would be even better to throw a warning to the user if unexpected legacy entries are found in descriptor wallets, but I think as a first mitigation everything is obvisouly better than crashing. As far as I'm aware, descriptor wallets created/migrated by Bitcoin Core should never end up in a state containing legacy type entries though.~~

This PR fixes this by throwing an error if legacy entries are found in descriptor wallets on loading.

achow101 commented:
Meh. I think it's reasonable to expect failures if the user is manually modifying their wallet file by doing direct database operations themselves. Since it is completely unexpected to have legacy and descriptor wallet things mixed, I don't think we should just silently ignore such issues and instead return an error.

theStack commented:
> Meh. I think it's reasonable to expect failures if the user is manually modifying their wallet file by doing direct database operations themselves. Since it is completely unexpected to have legacy and descriptor wallet things mixed, I don't think we should just silently ignore such issues and instead return an error.

Right, throwing an error seems more reasonable. Obviously it wouldn't be the user itself modifying an existing wallet, but e.g. an external malicious actor crafting and handing out these mixed wallets, hoping that users would open them to cause node crashes.

The PR in its current state is just the easiest "avoid crash"-remedy. Will look into your proposed
solution in a bit, this probably needs a new `DBErrors` category.

furszy commented:
Concept ACK in terms of not crashing, and instead abort cleanly with a message.

Even when this is an edge case (if people/attacker insert data into the wallet db, then the user might have worst problems then a simple crash..). Still, for those users experimenting with our wallet db; any crash could end up corrupting other dbs and indexes that the node has, which would require a fresh resync or reindex etc.. So, better if we don't crash and just abort cleanly.

Saying that, I would rather focus on getting #24914 merged first. The current load process is a mess, and continue adding stuff on top would just make it messier.

luke-jr commented:
Concept ACK on throwing an error.

>I would rather focus on getting https://github.com/bitcoin/bitcoin/pull/24914 merged first. The current load process is a mess, and continue adding stuff on top would just make it messier.

Bugfixes need to be backported, so this should ideally be merged first.

theStack committed:
wallet: throw error if legacy entries are present on loading descriptor wallets

In the wallet key-value-loading routine, most legacy type entries
require a LegacyScriptPubKeyMan instance after successful
deserialization. On a descriptor wallet, creating that (via method
`GetOrCreateLegacyScriptPubKeyMan`) fails and then leads to a
null-pointer dereference crash. Fix this by throwing an error if
if the wallet flags indicate that we have a descriptor wallet and there
is a legacy entry found.

theStack committed:
test: check that loading descriptor wallet with legacy entries throws error

theStack commented:
Force-pushed with a new fix that throws an error now, as suggested by reviewers (the old branch with the inferior ignore approach can still be found [here](https://github.com/theStack/bitcoin/tree/202211-wallet_fix_crash_on_descriptor_wallet_load_IGNORE)). Happy to improve the error message if there is any suggestions.

furszy commented:
> Concept ACK on throwing an error.
> 
> > I would rather focus on getting #24914 merged first. The current load process is a mess, and continue adding stuff on top would just make it messier.
> 
> Bugfixes need to be backported, so this should ideally be merged first.

True, still I'm not sure that adding a specific `DBError` is the right approach here. Some sort of db checksum created from one of the wallet encrypted keys might solve tampering attempts in a more general manner. 
This is because we most likely have other ways to crash/break the wallet at startup by inserting data directly into the db. Like modifying and/or erasing the wallet flags or the min version field, or changing a spkm type which would crash during the `GetNewDestination` type check assertion.

achow101 commented:
ACK 3198e4239e848bbb119e3638677aa9bcf8353ca6

aureleoules reviewed (approved):
ACK 3198e4239e848bbb119e3638677aa9bcf8353ca6

Tested with the steps in the PR description:

```sh
$ ./src/bitcoin-cli -regtest createwallet crashme
$ ./src/bitcoin-cli -regtest unloadwallet crashme
$ sqlite3 ~/.bitcoin/regtest/wallets/crashme/wallet.dat
sqlite> INSERT INTO main VALUES(x'07637363726970740000000000000000000000000000000000000000', x'00');

$ ./src/bitcoin-cli -regtest loadwallet crashme 
error code: -4
error message:
Wallet loading failed. Unexpected legacy entry in descriptor wallet found. Loading wallet /home/aureleoules/.bitcoin/regtest/wallets/crashme/wallet.dat

The wallet might have been tampered with or created with malicious intent.

```

aureleoules commented on src/wallet/walletdb.cpp:
```diff
@@ -833,6 +839,12 @@ DBErrors WalletBatch::LoadWallet(CWallet* pwallet)
             if (!ReadKeyValue(pwallet, ssKey, ssValue, wss, strType, strErr))
             {
+                if (wss.unexpected_legacy_entry) {
+                    strErr = strprintf("Error: Unexpected legacy entry found in descriptor wallet %s. ", pwallet->GetName());
+                    strErr += "The wallet might have been tampered with or created with malicious intent.";
+                    pwallet->WalletLogPrintf("%s\n", strErr);
```
I don't think there is a need to store the error in strErr?
```suggestion
                    pwallet->WalletLogPrintf("Error: Unexpected legacy entry found in descriptor wallet %s. The wallet might have been tampered with or created with malicious intent.\n", pwallet->GetName());
```
